stdlib/float.mli

```diff
@@ -257,11 +257,264 @@ val hash: t -> int
+      designate a valid subarray of [v2]. *)
+
+  val to_list : t -> float list
+  (** [to_list a] returns the list of all the elements of [a]. *)
+
+  val of_list : float list -> t
+  (** [of_list l] returns a fresh floatarray containing the elements
+      of [l].
+
+      Raise [Invalid_argument] if the length of [l] is greater than
+      [Sys.max_floatarray_length].*)
+
+  (* CR doligez: Should we insert subtitles as in [array.mli] ? *)
+
+  val iter : (float -> unit) -> t -> unit
+  (** [iter f a] applies function [f] in turn to all
+      the elements of [a].  It is equivalent to
+      [f a.(0); f a.(1); ...; f a.(length a - 1); ()]. *)
+
+  val iteri : (int -> float -> unit) -> t -> unit
+  (** Same as {!iter}, but the
+      function is applied with the index of the element as first argument,
+      and the element itself as second argument. *)
+
+  val map : (float -> float) -> t -> t
+  (** [map f a] applies function [f] to all the elements of [a],
+      and builds a floatarray with the results returned by [f]. *)
+
+  val mapi : (int -> float -> float) -> t -> t
+  (** Same as {!map}, but the
+      function is applied to the index of the element as first argument,
+      and the element itself as second argument. *)
+
+  val fold_left : ('a -> float -> 'a) -> 'a -> t -> 'a
+  (** [fold_left f x a] computes
+      [f (... (f (f x a.(0)) a.(1)) ...) a.(n-1)],
+      where [n] is the length of the floatarray [a]. *)
+
+  val fold_right : (float -> 'a -> 'a) -> t -> 'a -> 'a
+  (** [fold_right f a x] computes
+      [f a.(0) (f a.(1) ( ... (f a.(n-1) x) ...))],
+      where [n] is the length of the floatarray [a]. *)
+
+  val iter2 : (float -> float -> unit) -> t -> t -> unit
+  (** [Array.iter2 f a b] applies function [f] to all the elements of [a]
+      and [b].
+      Raise [Invalid_argument] if the floatarrays are not the same size. *)
+
+  val map2 : (float -> float -> float) -> t -> t -> t
+  (** [map2 f a b] applies function [f] to all the elements of [a]
+      and [b], and builds a floatarray with the results returned by [f]:
+      [[| f a.(0) b.(0); ...; f a.(length a - 1) b.(length b - 1)|]].
+      Raise [Invalid_argument] if the floatarrays are not the same size. *)
+
+  val for_all : (float -> bool) -> t -> bool
+  (** [for_all p [|a1; ...; an|]] checks if all elements of the floatarray
+      satisfy the predicate [p]. That is, it returns
+      [(p a1) && (p a2) && ... && (p an)]. *)
+
+  val exists : (float -> bool) -> t -> bool
+  (** [exists p [|a1; ...; an|]] checks if at least one element of
+      the floatarray satisfies the predicate [p]. That is, it returns
+      [(p a1) || (p a2) || ... || (p an)]. *)
+
+  val mem : float -> t -> bool
+  (** [mem a l] is true if and only if [a] is equal (in the sense of generic
+      equality) to an element of [l]. *)
+
+  val memq : float -> t -> bool
+  (** Same as {!mem}, but uses float equality instead of generic
+      equality to compare elements. *)
```

Comment:
The doc for mem and memq is confusing (or it is me being confused, perhaps).  "float equality" could be interpreted as Float.equal, which behaves as the generic comparison (i.e. Stdlib.compare x y = 0), but memq rather uses the generic equality (i.e. (=)).

Reply:
I'm not sure we're in a good state at the moment with regard to IEEE semantics on float equality.  I tend to think that it's a mistake that `Float.equal` does not have the IEEE semantics---my guess is that users would expect it to have such semantics (I certainly did), and indeed the polymorphic equality operator has such semantics.  I wonder if we should change `Float.equal` and add a second `Float.equal_non_ieee` or similar.  I know this means that `equal` functions cannot in general be defined as `compare x y = 0`, but I think that is an inevitable consequence if assuming IEEE semantics globally.